### PR TITLE
Add whereabouts package

### LINKED ISSUE #2370: Add whereabouts package
## Feature Request

TCE has `multus-cni` package which can help users to setup multiple networks for a pod. And this also requires an IPAM cni plugin to work together. However, TCE now has no package for IPAM cni plugin. So by default users can only use the IPAM plugin from k8s /opt/cni/bin, such as dhcp plugin, which cannot satisfy some requirements，like [issue](https://github.com/containernetworking/plugins/issues/640) or IPv6.

[whereabouts](https://github.com/k8snetworkplumbingwg/whereabouts) is a tool that assigns IP addresses cluster-wide and it supports IPv4&IPv6. This is also a preferred choice of some multus users. So It would be good if TCE support whereabouts package.

## Additional context

NetworkAttachmentDefinition with whereabouts works as below 
```
apiVersion: "k8s.cni.cncf.io/v1"
kind: NetworkAttachmentDefinition
metadata:
  name: macvlan-conf-1
spec:
  config: '{
      "cniVersion": "0.3.0",
      "name": "macvlan-conf-1",
      "type": "macvlan",
      "master": "eth0",
      "mode": "bridge",
      "ipam": {
        "type": "whereabouts",
        "range": "2001::0/116",
        "gateway": "2001::f:1",
        "range_start": "2001::f:10",
        "range_end": "2001::f:20"
      }
    }'
```

### PR DESCRIPTION
**Pull request by zhanggbj**
## What this PR does / why we need it

Whereabouts is a CNI IPAM plugin that assigns IP addresses cluster-wide.
It can work with Multus CNI plugin to help to provide multiple
network for pods.

## Details for the Release Notes (PLEASE PROVIDE)

```release-note
Add whereabouts package which is a CNI IPAM plugin
```

## Which issue(s) this PR fixes

Fixes: #2370 

## Special notes for your reviewer

- [x] create a package
- [x] Function test on TCE

### CODE REVIEW

**Review by seemiller — CHANGES_REQUESTED** [review]
Make sure that `hack/check-mdlint.sh` passes (there are a couple errors in the README)

A bigger problem is that the whereabouts uses an Alpine base image. Due to licensing issues, we cannot approve this package as is. I'll have to get back to you with steps to replace the Alpine image with something else.

File: addons/packages/whereabouts/0.4.2/README.md [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+## Usage Example
+
+The follow is a basic guide for getting started with whereabouts.
+
+This example guides you about attaching the second network interface to a pod with IP address assigned in the range you specified using whereabouts. 
+
+1. Install TCE Multus CNI package to support multiple network by following
+   [doc](https://github.com/vmware-tanzu/community-edition/blob/main/addons/packages/multus-cni/3.7.1/README.md#usage-example):
+
+1. Install TCE whereabouts package through tanzu CLI
```
[the remark this answers is not in the document]
Comment (zhanggbj) [reply addons/packages/whereabouts/0.4.2/README.md:40]: Fixed

**Comment by zhanggbj** [comment]
Hi @seemiller,

I've rebuilt the image based on `distroless` image and refined the whereabouts package, would you please to take a review, thanks!

(Housekeeping check is failing but it seems a common issue for git action. )

**Comment by zhanggbj** [comment]
This is ready for review, but I added the `hold` label as I'm contacting CNI community to publish a new release which includes many fixes recently. So we don't need to bump again in the short future.

**Comment by zhanggbj** [comment]
Verified installation and pod creatation.

```
$ tanzu package install whereabouts --package-name whereabouts.community.tanzu.vmware.com --version 0.5.0
- Installing package 'whereabouts.community.tanzu.vmware.com'
- Getting namespace 'default'
| Getting package metadata for 'whereabouts.community.tanzu.vmware.com'
- Creating service account 'whereabouts-default-sa'
- Creating cluster admin role 'whereabouts-default-cluster-role'
- Creating cluster role binding 'whereabouts-default-cluster-rolebinding'
- Creating package resource
| Package install status: Reconciling

 Added installed package 'whereabouts' in namespace 'default'

```

File: addons/packages/whereabouts/0.5.0/repo/.imgpkg/images.yml [hunk 2]
```diff
@@ -0,0 +1,7 @@
+---
```
Comment (chenliu1993) [inline addons/packages/whereabouts/0.5.0/repo/.imgpkg/images.yml:1]: What is the usage of this `repo` folder?

File: addons/packages/whereabouts/0.5.0/bundle/config/overlay/overlay-daemonset-install.yaml [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+kind: ServiceAccount
+metadata:
+  name: whereabouts
+  namespace: #@ data.values.namespace
+
+#@overlay/match by=overlay.subset({"kind":"ClusterRoleBinding"})
+---
+apiVersion: rbac.authorization.k8s.io/v1
+subjects:
+  #@overlay/match by="name"
```
Comment (chenliu1993) [inline addons/packages/whereabouts/0.5.0/bundle/config/overlay/overlay-daemonset-install.yaml:16]: Just a small thing, maybe we use the match in form of `key:value` to make it more specific. But I think it is ok if this match also works.

like #@overlay/match by=overlay.subset({"name": "whereabouts"})

File: addons/packages/whereabouts/0.5.0/bundle/config/overlay/overlay-daemonset-install.yaml [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+apiVersion: apps/v1
+kind: DaemonSet
+metadata:
+  name: whereabouts
+  namespace: #@ data.values.namespace
+spec:
+  template:
+    spec:
+      containers:
+        #@overlay/match by="name"
```
Comment (chenliu1993) [inline addons/packages/whereabouts/0.5.0/bundle/config/overlay/overlay-daemonset-install.yaml:32]: Same as above. 
Like #@overlay/match by=overlay.subset({"name": "whereabouts"})

File: addons/packages/whereabouts/0.5.0/README.md [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+
+This example guides you about attaching the second network interface to a pod with IP address assigned in the range you specified using whereabouts.
+
+1. Install TCE Multus CNI package to support multiple network by following
+   [doc](https://github.com/vmware-tanzu/community-edition/blob/main/addons/packages/multus-cni/3.7.1/README.md#usage-example):
+
+1. Install TCE whereabouts package through Tanzu CLI
+
+    ```bash
+    tanzu package install whereabouts --package-name whereabouts.community.tanzu.vmware.com --version ${MULTUS_PACKAGE_VERSION}
```
Comment (chenliu1993) [inline addons/packages/whereabouts/0.5.0/README.md:47]: Shoudl the version field be the whereabouts version?

File: addons/packages/whereabouts/0.5.0/README.md [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+    ```
+
+    > You can get the `${MULTUS_PACKAGE_VERSION}` from running `tanzu package
```
Comment (chenliu1993) [inline addons/packages/whereabouts/0.5.0/README.md:50]: Same as above

File: addons/packages/whereabouts/0.5.0/README.md [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+
+| Value | Required/Optional | Description |
+|-------|-------------------|-------------|
+| `namespace` | Optional | The namespace in which to deploy whereabouts components. Default: kube-system |
+
+### whereabouts Configuration
+
+| Value | Required/Optional | Description |
+|-------|-------------------|-------------|
+| `whereabouts.config.resources.limits.cpu` | Optional | The limits for CPU resources of whereabouts DeamonSet  |
```
Comment (chenliu1993) [inline addons/packages/whereabouts/0.5.0/README.md:27]: please make sure if they are optional when providing a values file
For example, if there is no `whereabouts.config.resources.requests.memory`, what is the final content of the affected resources spec.

File: addons/packages/whereabouts/0.5.0/README.md [hunk 8]
```diff
@@ -0,0 +1,108 @@
+# whereabouts Package
```
Comment (chenliu1993) [inline addons/packages/whereabouts/0.5.0/README.md:1]: s/whereabouts/Whereabouts/

File: addons/packages/whereabouts/0.5.0/repo/.imgpkg/images.yml (hunk 2 above)
Comment (zhanggbj) [reply addons/packages/whereabouts/0.5.0/repo/.imgpkg/images.yml:1]: This is a file for testing, will remove it!

File: addons/packages/whereabouts/0.5.0/bundle/config/overlay/overlay-daemonset-install.yaml (hunk 3 above)
Comment (zhanggbj) [reply addons/packages/whereabouts/0.5.0/bundle/config/overlay/overlay-daemonset-install.yaml:16]: Sounds good, will update it.

File: addons/packages/whereabouts/0.5.0/README.md (hunk 5 above)
Comment (zhanggbj) [reply addons/packages/whereabouts/0.5.0/README.md:47]: Good catch, will update it.

File: addons/packages/whereabouts/0.5.0/README.md (hunk 7 above)
Comment (zhanggbj) [reply addons/packages/whereabouts/0.5.0/README.md:27]: My understanding is that if we are providing something in the `values.yaml`, that means we have set default values, so I set it to `Optional`.

Leave this comment here and see if there're other opinions.

**Comment by zhanggbj** [comment]
@chenliu1993 Thanks for your review! I just addressed all the comments.

**Review by chenliu1993 — APPROVED** [review]
/lgtm

File: addons/packages/whereabouts/0.5.0/README.md (hunk 7 above)
Comment (seemiller) [reply addons/packages/whereabouts/0.5.0/README.md:27]: Most importantly is that all configuration parameters are documented.

**Review by seemiller — APPROVED** [review]
LGTM

**Pull request merged**